File: go/hashmap/128_longest_consecutive.go
package hashmap

// longestConsecutive 最长连续序列
func longestConsecutive(nums []int) int {
	// METHOD1: 排序后遍历
	// METHOD2: HashMap 先去重，然后找到连续序列中的最小元素x
	// 把x + 1, x+2 依次递增，观察hashmap 中是否有元素

	hmap := make(map[int]bool)

	for _, v := range nums {
		hmap[v] = true
	}

	result := 0
	// 遍历hmap 先找到连续序列中的最小元素
	// k 是nums 中的元素
	for k, _ := range hmap {

		// 如果v-1 不在hmap 里面，说明v 可以是连续序列的开始的值
		// if _, ok := hmap[k-1]; !ok {
		// 或者下面的写法
		if !hmap[k-1] {

			curNum := k
			length := 1

			for hmap[curNum+1] {
				curNum++
				length++
			}

			if result < length {
				result = length
			}
		}
	}
	return result
}

File: rust/src/hashmap/_128_longest_consecutive.rs
/// 最长连续序列
pub fn longest_consecutive(nums: Vec<i32>) -> i32 {
    // 先用hashMap 去重，
    // 找到连续序列中的最小值v(v-1 不在hmap 里面则是最小值)
    // 循环 v+1/v+2/... 查看是否在hmap 中

    use std::collections::HashSet;

    let mut set = HashSet::new();

    nums.iter().for_each(|&v| {
        set.insert(v);
    });

    let mut result = 0;
    set.iter().for_each(|&x| {
        if !set.contains(&(x - 1)) {
            let mut cur_num = x;
            let mut length = 1;

            while set.contains(&(cur_num + 1)) {
                cur_num += 1;
                length += 1;
            }

            if length > result {
                result = length;
            }
        }
    });

    result
}
